### PR TITLE
Eng 9460 dockerfile

### PR DESCRIPTION
**Pull request by jacobweinstock**
## Description

Add Dockerfile. Add port option for server subcommand. Updated Readme.

## Why is this needed

Fixes: #

## How Has This Been Tested?
Yes, manually. I ran `make image` and `make run-image` and then manually hit the API to test.

## How are existing users impacted? What migration steps/scripts do we need?

no, net new.

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.15 as builder
+
+COPY go.mod go.sum /
+RUN go mod download
+
+ADD . /code
```
Comment (mmlb) [inline Dockerfile:8]: Should be `COPY`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+WORKDIR /code
+RUN make build
+
+FROM scratch
+COPY --from=builder /code/bin/pbnj-linux-amd64 /pbnj-linux-amd64
+ENTRYPOINT ["/pbnj-linux-amd64", "server"]
```
Comment (mmlb) [inline Dockerfile]: missing new line

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.15 as builder
+
+COPY go.mod go.sum /
+RUN go mod download
+
+ADD . /code
+WORKDIR /code
+RUN make build
+
+FROM scratch
```
Comment (mmlb) [inline Dockerfile:11]: Lets do alpine. Nothing worse than trying to jump into a container to debug and run tcpdump or w/e and its based off of scratch!

File: Dockerfile (hunk 2 above)
Comment (mmlb) [inline Dockerfile]: Lets throw this up at the top, ENTRYPOINT is really just a container runtime metadata data field and I think it looks nice to have:

```dockerfile
FROM

EXPOSE
ENTRYPOINT
CMD

...
```

**Review by mmlb — CHANGES_REQUESTED** [review]
Lets split the README change away from the Dockerfile commit.
If we go with and alpine image we can simplify the runtime image dockerfile bits quite a bit and imo is better for production.

File: Dockerfile (hunk 1 above)
Comment (jacobweinstock) [reply Dockerfile:8]: ah yes, will update.

File: Dockerfile (hunk 2 above)
Comment (jacobweinstock) [reply Dockerfile]: is this organization from some standard or doc somewhere? I typically like to follow others who do more of a logical mind mapping for organizing from top to bottom. for example https://github.com/jessfraz/dockerfiles/blob/master/cfssl/Dockerfile

File: Dockerfile (hunk 3 above)
Comment (jacobweinstock) [reply Dockerfile:11]: hmm...a security first argument is that troubleshooting should be done by adding another container in the same namespace with tools like tcpdump/etc to do this type of troubleshooting.

File: Dockerfile (hunk 2 above)
Comment (jacobweinstock) [reply Dockerfile]: will update, thanks

Comment (jacobweinstock) [reply Dockerfile]: oh wait, this is an old file.

**Comment by jacobweinstock** [comment]
> Lets split the README change away from the Dockerfile commit.
> If we go with and alpine image we can simplify the runtime image dockerfile bits quite a bit and imo is better for production.

Readme removed

File: Dockerfile (hunk 2 above)
Comment (mikemrm) [reply Dockerfile]: Yeah I feel the same as @jacobweinstock when I look at a dockerfile I typically see entrypoint and cmd set at the bottom. It also just feels more proper that at the end you would define those.

**Review by mikemrm — COMMENTED** [review]
This looks good! I'm not sure if the `remove README update` request was asking to actually remove it from the pr, but rather just split it into a separate commit. I definitely think these changes should be in here.

**Comment by jacobweinstock** [comment]
> This looks good! I'm not sure if the `remove README update` request was asking to actually remove it from the pr, but rather just split it into a separate commit. I definitely think these changes should be in here.

ah ok, thanks for clarify that. I will add it back.

File: Dockerfile (hunk 2 above)
Comment (mmlb) [reply Dockerfile]: Just matches all the other repos in tinkerbell org, likely originating from me :D. My reasoning for it is its container metadata and not anything that changes the files on the layers. I like to have all the metadata up front and then can see how layers and files are going to be affected. Probably also busted the cache back in the day and likely not an issue if using buildx now adays.

Having one way of doing things (even if wrong) is better than multiple, so if we go with at the end (sounds fine to me), can you open up PRs for the other repos too please.

**Comment by mmlb** [comment]
> This looks good! I'm not sure if the `remove README update` request was asking to actually remove it from the pr, but rather just split it into a separate commit. I definitely think these changes should be in here.

yep

File: Dockerfile (hunk 3 above)
Comment (mmlb) [reply Dockerfile:11]: The thing is we don't have an easy way to attach containers to namespaces in the non-k8s sites. I think once we've transitioned to all k8s it makes sense to go from `scratch`

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+WORKDIR /code
+COPY go.mod go.sum /code/
+RUN go mod download
+
+COPY . /code
+RUN make build
+
+FROM scratch
+ARG DEFAULT_PORT=9090
+ENV PBNJ_PORT=${DEFAULT_PORT}
```
Comment (mmlb) [inline Dockerfile:13]: Why would we want an arg for default port? Seems a little weird to me. Changing the port can be done at run time via env and the arg isn't really used in the build process so doesn't really need to change there.

File: cmd/server.go [hunk 5]
```diff
@@ -22,63 +22,67 @@ const (
...
+				}
+				store, err := consul.NewClient(opts)
+				if err != nil {
+					logger.V(0).Error(err, "failed to connect to consul")
+					os.Exit(1)
+				}
+				var repo repository.Actions
+				repo = &persistence.GoKV{Store: store, Ctx: ctx}
+			*/
+			// add grpcsvr.WithPersistence(repo) to grpc.RunServer
```
Comment (mmlb) [inline cmd/server.go:75]: delete?

**Review by mmlb — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (jacobweinstock) [reply Dockerfile:11]: for non-k8s platforms, the pattern is to create the initial deployment manifest with a troubleshooting container.

File: Dockerfile (hunk 4 above)
Comment (jacobweinstock) [reply Dockerfile:13]: sure, I'll remove this.

File: cmd/server.go (hunk 5 above)
Comment (jacobweinstock) [reply cmd/server.go:75]: ah yes, will remove. thanks

Comment (jacobweinstock) [reply cmd/server.go:75]: actually, this is being removed in a different PR.

Comment (mmlb) [reply cmd/server.go:75]: heh :D, I thought it looked familiar. Lets ignore it here then.

File: Dockerfile (hunk 3 above)
Comment (mmlb) [reply Dockerfile:11]: sgtm

File: Makefile [hunk 6]
```diff
@@ -70,3 +70,11 @@ endif
...
 pbs: ## generate go stubs from protocol buffers
 	scripts/protoc.sh
+
+.PHONY: image
+image: ## make the Container Image
+	docker build -t pbnj:local . 
+
+.PHONY: run-image
+run-image: ## run PBnJ container image
+	docker run -it --rm -e PNBJ_LOGLEVEL=debug -p 9090:9090 pbnj:local
```
Comment (mmlb) [inline Makefile:80]: newline

File: scripts/etc-passwd [hunk 7]
```diff
@@ -0,0 +1 @@
+pbnj:x:10001:10001::/home/pbnj:/usr/sbin/nologin
```
Comment (mmlb) [inline scripts/etc-passwd:1]: newline

**Review by mmlb — CHANGES_REQUESTED** [review]

File: scripts/etc-passwd (hunk 7 above)
Comment (jacobweinstock) [reply scripts/etc-passwd:1]: just curious why? that I'm aware of, `/etc/passwd` files don't typically have a final newline.

File: Dockerfile (hunk 2 above)
Comment (jacobweinstock) [reply Dockerfile]: sure

File: Makefile (hunk 6 above)
Comment (jacobweinstock) [reply Makefile]: sure, updated

**Comment by jacobweinstock** [comment]
@mikemrm @jgavinray @mmlb just a friendly ping :)

**Review by mikemrm — APPROVED** [review]

**Comment by jacobweinstock** [comment]
waiting for https://github.com/tinkerbell/pbnj/pull/21 to merge to fix `go vet` issues

**Review by jgavinray — APPROVED** [review]
LGTM

**Pull request merged**